Spigot-Server-Patches/0613-Fix-item-position-desync.patch

```diff
@@ -0,0 +1,58 @@
+through the encryption/decryption methods of the packet that causes the
+precision loss, which forces the server to lose the same precision as
+the client keeping them in sync.
+
+diff --git a/src/main/java/com/destroystokyo/paper/PaperConfig.java b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+index 6ab0a51b421dc696aec292f2db955311c9cc31e3..c70b80ab32d7bac66f69a08f25726e73114bb7fe 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperConfig.java
+@@ -462,4 +462,9 @@ public class PaperConfig {
+     private static void trackPluginScoreboards() {
+         trackPluginScoreboards = getBoolean("settings.track-plugin-scoreboards", false);
+     }
++
++    public static boolean fixItemPositionDesync = true;
++    private static void fixItemPositionDesync() {
++        fixItemPositionDesync = getBoolean("settings.fix-item-position-desync", fixItemPositionDesync);
++    }
+ }
+diff --git a/src/main/java/net/minecraft/server/EntityTrackerEntry.java b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+index 4efc40c01ec12b80bd7cf9d35cf0ea0df973baf7..9c5a331aaa9d97ce1b8ea748705db5581ded4010 100644
+--- a/src/main/java/net/minecraft/server/EntityTrackerEntry.java
++++ b/src/main/java/net/minecraft/server/EntityTrackerEntry.java
+@@ -133,6 +133,12 @@ public class EntityTrackerEntry {
+                 double vec3d_dz = this.tracker.locZ() - 2.44140625E-4D*(this.zLoc);
+                 boolean flag1 = (vec3d_dx * vec3d_dx + vec3d_dy * vec3d_dy + vec3d_dz * vec3d_dz) >= 7.62939453125E-6D;
+                 // Paper end - reduce allocation of Vec3D here
++                // Paper start - fix MC-4
++                if (com.destroystokyo.paper.PaperConfig.fixItemPositionDesync && this.tracker instanceof EntityItem) {
++                    Vec3D loc = PacketPlayOutEntity.decrypt(PacketPlayOutEntity.encrypt(tracker.locX()), PacketPlayOutEntity.encrypt(tracker.locY()), PacketPlayOutEntity.encrypt(tracker.locZ()));
++                    tracker.setPosition(loc.getX(), loc.getY(), loc.getZ());
++                }
++                // Paper end - fix MC-4
+                 Packet<?> packet1 = null;
+                 boolean flag2 = flag1 || this.tickCounter % 60 == 0;
+                 boolean flag3 = Math.abs(i - this.yRot) >= 1 || Math.abs(j - this.xRot) >= 1;
+diff --git a/src/main/java/net/minecraft/server/PacketPlayOutEntity.java b/src/main/java/net/minecraft/server/PacketPlayOutEntity.java
+index e5da2b19c1177ba7f88f0aaad9d810bb313ce67b..7209a53f1f278acf6e7246d0cba5b4d1687b9f7a 100644
+--- a/src/main/java/net/minecraft/server/PacketPlayOutEntity.java
++++ b/src/main/java/net/minecraft/server/PacketPlayOutEntity.java
+@@ -14,10 +14,12 @@ public class PacketPlayOutEntity implements Packet<PacketListenerPlayOut> {
+     protected boolean h;
+     protected boolean i;
+ 
++    public static long encrypt(double d) { return a(d); } // Paper - OBFHELPER
```

Comment:
these are encode/decode, encryption isn't right word here

I'm thinking encodeDouble and decodeVector for names

Reply:
Done